File: foliage-rpc/src/main/java/org/lavenderx/foliage/nettyrpc/client/RpcProxy.java
package org.lavenderx.foliage.nettyrpc.client;

import org.lavenderx.foliage.nettyrpc.protocol.RpcRequest;
import lombok.extern.slf4j.Slf4j;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.util.UUID;

@Slf4j
public class RpcProxy<T> implements InvocationHandler {

    private final Class<T> clazz;

    public RpcProxy(Class<T> clazz) {
        this.clazz = clazz;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        if (Object.class == method.getDeclaringClass()) {
            String name = method.getName();
            if ("equals".equals(name)) {
                return proxy == args[0];
            } else if ("hashCode".equals(name)) {
                return System.identityHashCode(proxy);
            } else if ("toString".equals(name)) {
                return proxy.getClass().getName() + "@"
                        + Integer.toHexString(System.identityHashCode(proxy))
                        + ", with InvocationHandler " + this;
            } else {
                throw new IllegalStateException(String.valueOf(method));
            }
        }

        RpcRequest request = new RpcRequest();
        request.setRequestId(UUID.randomUUID().toString());
        request.setClassName(method.getDeclaringClass().getName());
        request.setMethodName(method.getName());
        request.setParameterTypes(method.getParameterTypes());
        request.setParameters(args);

        // Debug
        log.debug(method.getDeclaringClass().getName());
        log.debug(method.getName());
        for (int i = 0, l = method.getParameterTypes().length; i < l; ++i) {
            log.debug(method.getParameterTypes()[i].getName());
        }

        for (Object arg : args) {
            log.debug(arg.toString());
        }

        RpcClientHandler handler = ConnectionManager.getInstance().chooseHandler();
        RpcFuture rpcFuture = handler.sendRequest(request);

        return rpcFuture.get();
    }

    public RpcFuture call(String funcName, Object... args) {
        RpcClientHandler handler = ConnectionManager.getInstance().chooseHandler();
        RpcRequest request = createRequest(this.clazz.getName(), funcName, args);
        RpcFuture rpcFuture = handler.sendRequest(request);

        return rpcFuture;
    }

    private RpcRequest createRequest(String className, String methodName, Object[] args) {
        RpcRequest request = new RpcRequest();
        request.setRequestId(UUID.randomUUID().toString());
        request.setClassName(className);
        request.setMethodName(methodName);
        request.setParameters(args);

        Class[] parameterTypes = new Class[args.length];
        // Get the right class type
        for (int i = 0, l = args.length; i < l; i++) {
            parameterTypes[i] = getClassType(args[i]);
        }
        request.setParameterTypes(parameterTypes);

        log.debug(className);
        log.debug(methodName);
        for (Class parameterType : parameterTypes) {
            log.debug(parameterType.getName());
        }
        for (Object arg : args) {
            log.debug(arg.toString());
        }

        return request;
    }

    private Class<?> getClassType(Object obj) {
        Class<?> classType = obj.getClass();
        String typeName = classType.getName();
        switch (typeName) {
            case "java.lang.Integer":
                return Integer.TYPE;
            case "java.lang.Long":
                return Long.TYPE;
            case "java.lang.Float":
                return Float.TYPE;
            case "java.lang.Double":
                return Double.TYPE;
            case "java.lang.Character":
                return Character.TYPE;
            case "java.lang.Boolean":
                return Boolean.TYPE;
            case "java.lang.Short":
                return Short.TYPE;
            case "java.lang.Byte":
                return Byte.TYPE;
        }

        return classType;
    }

}

File: foliage-rpc/src/main/kotlin/org/lavenderx/foliage/nettyrpc/client/RpcProxy.kt
package org.lavenderx.foliage.nettyrpc.client

import org.lavenderx.foliage.nettyrpc.protocol.RpcRequest
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.util.*

class RpcProxy<T>(val clazz: Class<T>) : InvocationHandler {

    @Throws(Throwable::class)
    override fun invoke(proxy: Any, method: Method, args: Array<Any>): Any {
        if (proxy.javaClass == method.declaringClass) {
            val name = method.name
            if ("equals" == name) {
                return proxy === args[0]
            } else if ("hashCode" == name) {
                return System.identityHashCode(proxy)
            } else if ("toString" == name) {
                return proxy.javaClass.name + "@" + Integer.toHexString(System.identityHashCode(proxy)) + ", with InvocationHandler " + this
            } else {
                throw IllegalStateException(method.toString())
            }
        }

        val request = RpcRequest(UUID.randomUUID().toString())
        request.className = method.declaringClass.name
        request.methodName = method.name
        request.parameterTypes = method.parameterTypes
        request.parameters = args

        val handler = ConnectionManager.getInstance().chooseHandler()
        val rpcFuture = handler.sendRequest(request)

        return rpcFuture.get()!!
    }

    @Suppress("UNCHECKED_CAST")
    fun call(funcName: String, vararg args: Any): RpcFuture {
        val handler = ConnectionManager.getInstance().chooseHandler()
        val request = createRequest(this.clazz.name, funcName, args as Array<Any>)
        val rpcFuture = handler.sendRequest(request)

        return rpcFuture
    }

    private fun createRequest(className: String, methodName: String, args: Array<Any>): RpcRequest {
        val request = RpcRequest(UUID.randomUUID().toString())
        request.className = className
        request.methodName = methodName
        request.parameters = args

        val parameterTypes = arrayOfNulls<Class<*>>(args.size)
        // Get the right class type
        var i = 0
        val l = args.size
        while (i < l) {
            parameterTypes[i] = getClassType(args[i])
            i++
        }
        request.parameterTypes = parameterTypes.filterNotNull().toTypedArray()

        return request
    }

    private fun getClassType(obj: Any): Class<*> {
        val classType = obj.javaClass
        val typeNamedClass = when (classType.name) {
            "java.lang.Integer" -> Integer.TYPE
            "java.lang.Long" -> java.lang.Long.TYPE
            "java.lang.Float" -> java.lang.Float.TYPE
            "java.lang.Double" -> java.lang.Double.TYPE
            "java.lang.Character" -> java.lang.Character.TYPE
            "java.lang.Boolean" -> java.lang.Boolean.TYPE
            "java.lang.Short" -> java.lang.Short.TYPE
            "java.lang.Byte" -> java.lang.Byte.TYPE
            "kotlin.Int" -> kotlin.Int::class.java
            "kotlin.Long" -> kotlin.Long::class.java
            "kotlin.Float" -> kotlin.Float::class.java
            "kotlin.Double" -> kotlin.Double::class.java
            "kotlin.Char" -> kotlin.Char::class.java
            "kotlin.Boolean" -> kotlin.Boolean::class.java
            "kotlin.Short" -> kotlin.Short::class.java
            "kotlin.Byte" -> kotlin.Byte::class.java
            else -> null
        }

        if (typeNamedClass != null) {
            return typeNamedClass
        } else {
            return classType
        }
    }
}